### PR TITLE
support when closure

### LINKED ISSUE #3986: support VS Code `when` closure contexts
https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts

### LINKED ISSUE #4077: [plugin] control command enablement based on `menus.commandPalette` contribution point
https://code.visualstudio.com/api/extension-guides/command#controlling-when-a-command-shows-up-in-the-command-palette

### PR DESCRIPTION
**Pull request by akosyakov**
fix #3986, fix #4077

This PR:
-  properly hooks up Monaco context key service with the core and plugin system services;
- introduce [default VS Code context keys](https://github.com/theia-ide/theia/issues/3986#issuecomment-454364740);
- manage commands in the command palette based on `menus.commandPalette` contribution point (https://github.com/theia-ide/theia/issues/4077) and when context;
- preserver focus when quick open is opened recursively (#4098) - needed to ensure that values are computed in the proper context (active html element);
- support of configuration settings contexts https://github.com/theia-ide/theia/issues/4073

#### Out of scope
- Deprecate existing keybinding context and replace it with when closure. https://github.com/theia-ide/theia/issues/4108
- Support of active/focused view or panel contexts https://github.com/theia-ide/theia/issues/4104

#### Testing

VS Code extension to test: https://github.com/akosyakov/vscode-when-closure

How:
- clone it in theia/plugins folder
- start browser example
- try commands in different contexts: https://github.com/akosyakov/vscode-when-closure/blob/master/package.json 
- compare availability of commands with VS Code

Contexts:
- [x] editor
- [x] os
- [x] mode
  - [x] debug
  - [x] snippet
  - [x] quick open
- [x] resource
- [x] explorer
- [x] editor widget
- [x] terminal
- [x] global UI
  - [x] notification
  - [x] search
  - [x] shell
  - [x] console (debug)
  - [x] editors
  - [x] workspace
- [x] configuration settings

### CODE REVIEW

File: packages/core/src/browser/context-key-service.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (C) 2018 Red Hat, Inc. and others.
+ * Copyright (C) 2019 TypeFox and others.
```
Comment (akosyakov) [inline packages/core/src/browser/context-key-service.ts]: it is GH glitch, i did not copy it

File: packages/search-in-workspace/src/browser/search-in-workspace-context-key-service.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import { injectable, inject, postConstruct } from 'inversify';
+import { ContextKeyService, ContextKey } from '@theia/core/lib/browser/context-key-service';
+
+@injectable()
+export class SearchInWorkspaceContextKeyService {
+
+    @inject(ContextKeyService)
+    protected readonly contextKeyService: ContextKeyService;
+
+    protected _searchViewletVisible: ContextKey<boolean>;
```
Comment (benoitf) [inline packages/search-in-workspace/src/browser/search-in-workspace-context-key-service.ts]: I don't know if it was clarified last week at the meeting but AFAIK underscore shouldn't be used per coding rules

**Comment by akosyakov** [comment]
@azatsarynnyy If you are available it would be nice if you can test `view` and `viewItem` context keys. I've removed copied `ContextKeyService` from VS Code and instead exposed it from Monaco.

**Comment by akosyakov** [comment]
In addition to testing with [vscode-when-closure](vscode-when-closure), one can pack and download `emmet` extension from https://github.com/theia-ide/vscode-builtin-extensions. Emmet commands should be enabled only in the file resource context: https://github.com/Microsoft/vscode/blob/f20fb4e7c907a5835f4bc9c5680614220efabf7c/extensions/emmet/package.json#L340

**Comment by svenefftinge** [comment]
When we register the monaco keybindings on our registry we should use the 'when' context instead of our context. I created a ticket as a follow up to this PR: https://github.com/theia-ide/theia/issues/4114

File: packages/debug/src/browser/debug-session-manager.ts [hunk 3]
```diff
@@ -105,11 +107,23 @@ export class DebugSessionManager {
...
 
+    @inject(ContextKeyService)
+    protected readonly contextKeyService: ContextKeyService;
+
+    protected debugTypeKey: ContextKey<string>;
+    protected inDebugModeKey: ContextKey<boolean>;
+
     @postConstruct()
     protected init(): void {
+        this.debugTypeKey = this.contextKeyService.createKey<string>('debugType', undefined);
```
Comment (svenefftinge) [inline packages/debug/src/browser/debug-session-manager.ts]: couldn't find this context in https://github.com/Microsoft/vscode-docs/blob/master/docs/getstarted/keybindings.md#when-clause-contexts

**Comment by azatsarynnyy** [comment]
@akosyakov my vacation ends on Jan 21, so I'll test it on Tue

**Comment by akosyakov** [comment]
@azatsarynnyy Would you have a chance to test?

**Comment by azatsarynnyy** [comment]
@akosyakov I'm testing it now

**Comment by azatsarynnyy** [comment]
@akosyakov
it appears that `menus` contribution point stopped working. I've investigated a bit and find out that the changes in `command-registry-main.ts` from that PR https://github.com/theia-ide/theia/commit/0ce38188ed3010662b90feacd3097d1b7eeb36c1#diff-fd4fb4ae66141acec5a29ca31770d486L41 also should be reverted to fix that

**Comment by akosyakov** [comment]
@azatsarynnyy thanks for checking, can you clarify how you test it, i want to reproduce it before

**Comment by azatsarynnyy** [comment]
@akosyakov I'm using the `tree-view-sample-plugin` from the repo with plugin samples https://github.com/eclipse/che-theia-samples/tree/when
I've pushed my changes for testing into `when` branch.
To test it against the locally modified Theia I'm using the hosted mode:
- open the repo in Theia workspace
- use `Hosted Plugin: Start Instance` command
- choose `tree-view-sample-plugin` folder
- in opened tab with development host open `Comments` view and there should be context menus for the tree items

**Comment by azatsarynnyy** [comment]
It seems that `config.*` context keys also stopped working

**Comment by akosyakov** [comment]
> It seems that config.* context keys also stopped working

@azatsarynnyy could you point out where it was hooked? I would try to restore it. There is https://github.com/theia-ide/theia/issues/4073 to properly hook up Monaco with Theia preference service then `config.*` would work properly for Monaco and plugins. But it would make this PR humongous, was planning to tackle it separately.

**Comment by azatsarynnyy** [comment]
@akosyakov it was hooked here https://github.com/theia-ide/theia/pull/3879/commits/aedd3aaf8f80c4946b05b075fe1badf33c4313da#diff-7dbe1e9893e323fbe9847973ce87d3ebR111

**Comment by akosyakov** [comment]
@azatsarynnyy I could not find `menu` contribution point in this branch: https://github.com/eclipse/che-theia-samples/tree/when I'm not sure how it is supposed to work.

**Comment by azatsarynnyy** [comment]
@akosyakov it's here https://github.com/eclipse/che-theia-samples/blob/when/tree-view-sample-plugin/package.json#L45-L60

File: packages/monaco/package.json [hunk 4]
```diff
@@ -10,6 +10,7 @@
     "@theia/markers": "^0.3.19",
     "@theia/outline-view": "^0.3.19",
     "@theia/workspace": "^0.3.19",
+    "deepmerge": "2.0.1",
```
Comment (akosyakov) [inline packages/monaco/package.json]: @marcdumais-work we already have deepmerge in yarn.lock via wdio dev dependency, do i need to open another CQ?

**Comment by akosyakov** [comment]
@azatsarynnyy I've addressed your issues. Could you try again please?

**Comment by akosyakov** [comment]
Now only CQ hell 🙄

File: packages/monaco/package.json (hunk 4 above)
Comment (marcdumais-work) [reply packages/monaco/package.json]: Hi @akosyakov,

Good news: no need for a new or updated CQ: "deepmerge": "2.0.1" was already covered in recent [CQ 18717](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=18717)

**Comment by akosyakov** [comment]
@benoitf @evidolob @AlexTugarev objections to merge?

File: packages/monaco/package.json (hunk 4 above)
Comment (marcdumais-work) [reply packages/monaco/package.json]: We should probably have waited until the CQ above was approved or given "merge permission" before merging this PR. In theory if this new dependency is not license-compatible, we would need to take action to remove it from the repo... (should be ok in this specific case though)

Comment (marcdumais-work) [reply packages/monaco/package.json]: Next time I'll add a review comment on the PR to reflect this situation.

**Pull request merged**